### ISSUE TITLE
minimum GPU memory?

### ISSUE THREAD

**Issue by Skwarson96** [InstantID/InstantID#61]
Hello
What is minimum GPU memory value required to run the code (on default settings/models from readme)?

**Comment by johndpope** [comment]
should match sdxl 
6GB VRAM – SDXL will work better than on a 4GB card, but it’s still not enough for comfortable work. Would you like to wait up to an hour to generate one 1024×1024 image?

**Comment by Skwarson96** [comment]
Nah, I close issue to fast. I see these is related issue: https://github.com/InstantID/InstantID/issues/72

**Comment by rupeshs** [comment]
@Skwarson96 https://github.com/InstantID/InstantID/issues/21

**Comment by mwmw7** [comment]
It is working on A5000 now

**Comment by hortom** [comment]
It sometimes can run on an RTX 4070 with 12 GB.

**Comment by KurtCocain** [comment]
> Rarely, it can run on an RTX 4070 with 12 GB.

How? I have a 12gb 4070ti and it does not run. Even with everything closed, just launching it takes 11.8gb of my ram, starting a render OOM's.

Do you have a specific setup that uses less memory?

**Comment by bfrye26** [comment]
> > Rarely, it can run on an RTX 4070 with 12 GB.
> 
> How? I have a 12gb 4070ti and it does not run. Even with everything closed, just launching it takes 11.8gb of my ram, starting a render OOM's.
> 
> Do you have a specific setup that uses less memory?

Also curious how? I am running a 4080 and it still hits the memory limit, and no idea why. If people can run it with under 10GB something must be wrong but no idea how to find it.

**Comment by b4sh** [comment]
> > Rarely, it can run on an RTX 4070 with 12 GB.
> 
> How? I have a 12gb 4070ti and it does not run. Even with everything closed, just launching it takes 11.8gb of my ram, starting a render OOM's.
> 
> Do you have a specific setup that uses less memory?

I run this on RTX4070. 1024x1024 image generate in ~60sec. But modifications is needed (xformers optimization). I run this on Linux running in WSL2. Image generation takes full VRAM and 4G shared memory. If the entire SDXL pipeline could fit into VRAM, then generating an image would take about 22 seconds.

**Comment by bfrye26** [comment]
> > > Rarely, it can run on an RTX 4070 with 12 GB.
> > 
> > 
> > How? I have a 12gb 4070ti and it does not run. Even with everything closed, just launching it takes 11.8gb of my ram, starting a render OOM's.
> > Do you have a specific setup that uses less memory?
> 
> I run this on RTX4070. 1024x1024 image generate in ~60sec. But modifications is needed (xformers optimization). I run this on Linux running in WSL2. Image generation takes full VRAM and 4G shared memory. If the entire SDXL pipeline could fit into VRAM, then generating an image would take about 22 seconds.

Could you share some of the optimizations you run?

**Comment by KurtCocain** [comment]
> > > > Rarely, it can run on an RTX 4070 with 12 GB.
> > > 
> > > 
> > > How? I have a 12gb 4070ti and it does not run. Even with everything closed, just launching it takes 11.8gb of my ram, starting a render OOM's.
> > > Do you have a specific setup that uses less memory?
> > 
> > 
> > I run this on RTX4070. 1024x1024 image generate in ~60sec. But modifications is needed (xformers optimization). I run this on Linux running in WSL2. Image generation takes full VRAM and 4G shared memory. If the entire SDXL pipeline could fit into VRAM, then generating an image would take about 22 seconds.
> 
> Could you share some of the optimizations you run?

+1 on that, I wouldn't know how to implement that myself. Also, i run windows

**Comment by b4sh** [comment]
> Could you share some of the optimizations you run?

https://huggingface.co/docs/diffusers/main/optimization/memory#memory-efficient-attention

**Comment by KurtCocain** [comment]
> > Could you share some of the optimizations you run?
> 
> https://huggingface.co/docs/diffusers/main/optimization/memory#memory-efficient-attention
> 
> And FaceAnalysis must be run on CPU - providers=['CPUExecutionProvider']

I'm sorry if this is ridiculously easy but I'm gonna need more information. 

For facenalysis, its already on providers=['CPUExecutionProvider'] by default in the app.py, i really have no idea what i'm supposed to look for

As for the memory efficient attention, where exactly am i supposed to import/pipe.enable it?

**Comment by b4sh** [comment]
I created fork with my mods. This is tested only on Linux (WSL2) and RTX4070.

https://github.com/b4sh/InstantID/tree/main

You need install everything from scratch. Clone my repo, make venv and install requirements from gradio_demo/requirements.txt

**Comment by hortom** [comment]
> > Rarely, it can run on an RTX 4070 with 12 GB.
> 
> How? I have a 12gb 4070ti and it does not run. Even with everything closed, just launching it takes 11.8gb of my ram, starting a render OOM's.
> 
> Do you have a specific setup that uses less memory?

I am using this [InstantID addon](https://github.com/ZHO-ZHO-ZHO/ComfyUI-InstantID) for [ComfyUI](https://github.com/comfyanonymous/ComfyUI). I guess that the latter one has some optimisation under the hood which allows InstantID to run mostly the first time then just OOM error.

**Comment by camoody1** [comment]
> Could you share some of the optimizations you run?

I'm pretty sure his biggest optimization is running on Linux. Most of the people I've seen be successful with InstantID are running Linux or a non-portable version of ComfyUI on Windows.

**Comment by camoody1** [comment]
> I created fork with my mods. This is tested only on Linux (WSL2) and RTX4070.
> 
> https://github.com/b4sh/InstantID/tree/main
> 
> You need install everything from scratch. Clone my repo, make venv and install requirements from gradio_demo/requirements.txt

I didn't reinstall everything from scratch, but I did take all of the changes that you made in your fork and add them to the current InstantID code I have installed. Most of your changes were just slider changes, but the:

_pipe.enable_xformers_memory_efficient_attention()_
....
_if torch.cuda.is_available():
   torch.cuda.empty_cache()_ 
... and ...
_def cuda(self, dtype=torch.float16, use_xformers=True):_

... changes seem to have made a nice little difference. Instead of taking several minutes to complete when it DOES work, now it only takes about 45 seconds. My VRAM seems to stay under the 12GB limit, now. Oddly, though, it only seems to work every other run. It runs once... fine. Next run... OOM. Next run... fine. Next run... OOM. So weird. But far better than it was. Thank you.

**Comment by camoody1** [comment]
@b4sh Man, I was just watching my Windows Task Manager as my workflow ran. It stayed juuuuust under 12GB all the way through until the very last step of creating the image. At that point, it jumped an additional 5.5GB!! And it all went onto my system RAM. That explains why that last step from 100% to image creation takes almost 10 seconds when the entire rest of it only takes 20 seconds.

What is kicking in that extra 5.5GB at the very end? Is there a VAE getting loaded in at the last minute? Can we clear any VRAM before that sucker hits?

**Comment by b4sh** [comment]
@camoody1 my every run takes about 60sec, no OOM error. And yes, this SDXL pipeline (SDXL checkpoint + IP_Adapter + ControlNet) need more than 12G VRAM, so part of the allocated memory is in the shared space. And there's not much that can be done about it. For this SDXL pipeline to operate quickly and efficiently, a card with 24GB of VRAM is needed.

**Comment by camoody1** [comment]
@b4sh Well then, there's only one thing left to ask... Would you please buy me an RTX 4090 24GB? I can give you $200 towards it. 😁
